### PR TITLE
Typescript

### PR DESCRIPTION
**Pull request by myme**
Moving this PR to the core repo. Planning to do some pre-releases, and don't want to do that from personal repos.

### CODE REVIEW

**Comment by myme** [comment]
Adding reference to original TypeScript issue:

Fix for #11.

File: src/connect.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * @param {string} [options.username] -
+ *     Username to authenticate with if protocol requires authentication.
+ * @param {string} [options.password] -
+ *     Password to authenticate with if protocol requires authentication.
+ * @param {string} [options.loglevel] -
+ *     Set the internal log level.
+ * @return {XAPI} - XAPI interface connected to the given URI.
+ */
+export default function connect(url: string, options: Options, initBackend: InitBackend) {
+  if (arguments.length === 1 && typeof url === 'object') {
```
Comment (bjornars) [inline src/connect.ts]: this doesnt typescript well at all, the arguments.length == 1 overload doesnt typecheck.

The shortest thing I could find that works is

```
interface Connect {
  (options: object): number;
  (url: string, options: object): number;
}

const connect: Connect = (...args: any[]) => {
  return args.length;
};

connect('str', {});
connect({});
```

File: src/index.ts [hunk 2]
```diff
@@ -57,62 +24,28 @@ function initBackend(opts: Options): Backend {
...
-  opts.host = opts.hostname;
-  delete opts.hostname;
-
-  log.setLevel(opts.loglevel);
-  log.info('connecting to', url);
-
-  const backend = initBackend(opts);
-  return new XAPI(backend);
+export function connect(url: string, options: Options): XAPI {
+  const opts = Object.assign({
```
Comment (bjornars) [inline src/index.ts]: splat this?

File: src/xapi/index.ts [hunk 3]
```diff
@@ -87,7 +94,7 @@ export default class XAPI extends EventEmitter {
 
   constructor(
     private readonly backend: Backend,
-    options: XapiOptions = {}) {
+    options: XapiOptions = { seal: true }) {
```
Comment (bjornars) [inline src/xapi/index.ts]: download (5)

File: src/cli.ts [hunk 4]
```diff
@@ -2,27 +2,28 @@
...
 
-import pkg from '../package.json';
+// this import messes up the package structure after running tsc
+// import * as pkg from '../package.json';
+import { connect } from './';
 import log from './log';
-import { connect } from '.';
+import XAPI from './xapi/index.js';
 
+const pkg = { version: '4.2.0' };
```
Comment (bjornars) [inline src/cli.ts]: This needs to be fixed somehow. We imported `package.json` before, but `tsc` was not the biggest fan.

**Pull request merged**